cirq-core/cirq/transformers/analytical_decompositions/quantum_shannon_decomposition.py

```diff
@@ -0,0 +1,248 @@
+# Copyright 2023 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+"""Utility methods for decomposing arbitrary n-qubit (2^n x 2^n) unitary.
+
+Based on:
+Synthesis of Quantum Logic Circuits. Tech. rep. 2006,
+https://arxiv.org/abs/quant-ph/0406176
+"""
+from typing import List, Iterator
+
+from scipy.linalg import cossin
+
+import numpy as np
+
+from cirq import ry, rz, CNOT, ZPowGate
+from cirq import is_unitary, deconstruct_single_qubit_matrix_into_angles, Operation
+from cirq.ops import Qid
+
+
+def quantum_shannon_decomposition(qubits: List[Qid], u: np.ndarray) -> List[Operation]:
```

Comment:
You can simply return a `cirq.OP_TREE` from `quantum_shannon_decomposition` instead of having a seperate `qsd_generator` method. 

`OP_TREE` is a recursive type defined as `OP_TREE = Union[Operation, Iterable['OP_TREE']]`. Users can call `cirq.flatten_op_tree(quantum_shannon_decomposition())` to get a `List[Operation]`  corresponding to the OP_TREE returned by your method. 

```suggestion
def quantum_shannon_decomposition(qubits: List[Qid], u: np.ndarray) -> cirq.OP_TREE:
```